cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/AbstractMapReduceContextBuilder.java

```diff
@@ -109,8 +112,8 @@ public BasicMapReduceContext build(MapReduceMetrics.TaskType type,
     MapReduceSpecification spec = program.getApplicationSpecification().getMapReduce().get(program.getName());
     BasicMapReduceContext context =
       new BasicMapReduceContext(program, type, RunIds.fromString(runId), taskId, runtimeArguments, datasets, spec,
-                                logicalStartTime, workflowBatch, discoveryServiceClient, metricsCollectionService,
-                                datasetFramework, adapterSpec, pluginInstantiator);
+                                logicalStartTime, workflowBatch, workflowToken, discoveryServiceClient,
```

Comment:
So for example when we run the `PurchaseHistoryWorkflow`, workflowBatch will be the `PurchaseHistoryBuilder`, the MapReduce program run by the Workflow. In `AbstractMapReduceContextBuilder`, while building the MapReduce context, if workflowBatch is not null, then we create the `Program` by fetching the specification for the MapReduce program from application specifications.
Corresponding code snippet in `AbstractMapReduceContextBuilder`:
```java
    // Initializing Program
    Program program = mrProgram;

    // See if it was launched from Workflow; if it was, change the Program.
    if (workflowBatch != null) {
      MapReduceSpecification mapReduceSpec = program.getApplicationSpecification().getMapReduce().get(workflowBatch);
      Preconditions.checkArgument(mapReduceSpec != null, "Cannot find MapReduceSpecification for %s", workflowBatch);
      program = new WorkflowMapReduceProgram(program, mapReduceSpec);
    }
```